File: src/main/js/structure/ConstantPoolSection.js
function ConstantPoolSection(elementCountVar, bytes, index) {

        var elementCount = elementCountVar;
//        var elementCount = 14;
        var startByteIndex = index;
        var length = 0;
        var constants = [];
        var selectedSection = null;
        var selectedIndex = null;

        // we will need the last one to determine the entire pool length
        var constantType = null;
        for(var i = 0; i < elementCount -1; i++) {
            var tag = bytes[index];
            constantType = createConstantType(tag, bytes, index);
            if(constantType != null) {
                constants.push(constantType);
                index += constantType.length();
            }
        }
        if (constantType != null) {
            length = constantType.getStartByteIndex() + constantType.length() - startByteIndex;
        }

    /**
     * How many elements in this section
     *
     * @return Number of elements in this section
     */
    this.elementCount = function() {
        return elementCount;
    }

    this.getSelectedIndex = function() {
        return selectedIndex;
    }

    /**
     * Index to the first byte of this section relative to the constant pool
     *
     * @return
     */
    this.getStartByteIndex = function() {
        return startByteIndex;
    }

    /**
     * length in bytes of this section
     *
     * @return length in bytes of this section
     */
    this.length = function() {
        return length;
    }

    this.getEndByteIndex = function () {
        return this.getStartByteIndex() + this.length();
    }

    this.contains = function (index) {
        if(index >= this.getStartByteIndex() && index < this.getStartByteIndex() + this.length()) {
            return true;
        }
        else {
            return false;
        }
    }

    this.getConstants = function() {
        return constants;
    }

    /**
     * Assuming the supplied index is the number of the constant (where the first is one) returns that constant
     */
    this.getConstant = function (baseOneIndex) {
        return constants[baseOneIndex -1];
    }


    this.getCssClass = function() {
        return "constant-pool-section";
    }

    /**
     * Index is the index into the byte array
     */
    this.getBytecode = function(index) {
        selectedSection = this.getSection(index);
        if(selectedSection != null) {
            return selectedIndex + " " + selectedSection.getBytecode(index, constants);
        }
    }

    this.getSection = function(index) {
        var section = null;
        constants.some(
            function(aSection, pointer) {

                if(aSection.contains(index)) {
                    section = aSection;
                    selectedIndex = pointer;
                    return true;
                }
                else {
                    return false;
                }
            }
        );
        return section;
    };
}

File: src/main/java/com/vgrazi/bytecodeexplorer/structure/ConstantPoolSection.java
package com.vgrazi.bytecodeexplorer.structure;

import com.vgrazi.bytecodeexplorer.structure.constantTypes.ConstantType;
import com.vgrazi.bytecodeexplorer.utils.ConstantTypeFactory;
import com.vgrazi.bytecodeexplorer.utils.Utils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by vgrazi on 8/13/15.
 */
public class ConstantPoolSection implements ClassFileSection {

    private int startByteIndex;
    private int length;

    private int elementCount;
    private final static List<ConstantType> constants = new ArrayList<>();

    public ConstantPoolSection(int elementCount, byte[] bytes, int index) {

        setElementCount(elementCount);
        this.startByteIndex = index;
        ConstantType constantType = null;
        for(int i = 0; i < elementCount -1; i++) {
            byte tag = bytes[index];
            constantType = ConstantTypeFactory.createConstantType(tag);
            constantType.setData(bytes, index);
            constants.add(constantType);
            index += constantType.length();
        }
        if (constantType != null) {
            length = constantType.getStartByteIndex() + constantType.length() - startByteIndex;
        }

        Utils.printConstants(constants);
    }

    public int getSectionIndex(int byteIndex) {
        for(int i = 0; i < constants.size(); i++) {
            ConstantType constant = constants.get(i);
            if(constant.contains(byteIndex)) {
                return i;
            }
        }
        return 0;
    }

    public void setElementCount(int elementCount) {
        this.elementCount = elementCount;
    }

    /**
     * length in bytes of this section
     *
     * @return length in bytes of this section
     */
    @Override
    public int length() {
        return length;
    }

    /**
     * Index to the first byte of this section relative to the constant pool
     *
     * @return
     */
    @Override
    public int getStartByteIndex() {
        return startByteIndex;
    }

    public ClassFileSection getSection(int sectionIndex) {
        ConstantType type = null;
        if (sectionIndex < constants.size()) {
            type = constants.get(sectionIndex);
        }
        return type;
    }

    public static List<ConstantType> getConstants() {
        return constants;
    }
}
